Source/WebKit/UIProcess/Extensions/WebExtension.cpp

```diff
@@ -237,6 +245,189 @@ void WebExtension::populateWebAccessibleResourcesIfNeeded()
         parseWebAccessibleResourcesVersion2();
 }
 
+static int toAPI(WebExtension::Error error)
+{
+    switch (error) {
+    case WebExtension::Error::Unknown:
+        return static_cast<int>(WebExtension::APIError::Unknown);
+    case WebExtension::Error::ResourceNotFound:
+        return static_cast<int>(WebExtension::APIError::ResourceNotFound);
+    case WebExtension::Error::InvalidManifest:
+        return static_cast<int>(WebExtension::APIError::InvalidManifest);
+    case WebExtension::Error::UnsupportedManifestVersion:
+        return static_cast<int>(WebExtension::APIError::UnsupportedManifestVersion);
+    case WebExtension::Error::InvalidDeclarativeNetRequest:
+        return static_cast<int>(WebExtension::APIError::InvalidDeclarativeNetRequestEntry);
+    case WebExtension::Error::InvalidBackgroundPersistence:
+        return static_cast<int>(WebExtension::APIError::InvalidBackgroundPersistence);
+    case WebExtension::Error::InvalidResourceCodeSignature:
+        return static_cast<int>(WebExtension::APIError::InvalidResourceCodeSignature);
+    default:
+        return static_cast<int>(WebExtension::APIError::InvalidManifestEntry);
+    }
+}
+
+Ref<API::Error> WebExtension::createError(Error error, const String& customLocalizedDescription, RefPtr<API::Error> underlyingError)
+{
+    auto errorCode = toAPI(error);
+    String localizedDescription;
+    RefPtr<JSON::Value> actionManifest;
+
+    switch (error) {
+    case Error::Unknown:
+        localizedDescription = WEB_UI_STRING("An unknown error has occurred.", "WKWebExtensionErrorUnknown description");
+        break;
+
+    case Error::ResourceNotFound:
+        ASSERT(customLocalizedDescription);
+        break;
+
+    case Error::InvalidManifest:
+        if (underlyingError && !underlyingError->localizedDescription().isEmpty())
+            localizedDescription = WEB_UI_FORMAT_STRING("Unable to parse manifest: %s", "WKWebExtensionErrorInvalidManifest description, because of a JSON error", underlyingError->localizedDescription().utf8().data());
+        else
+            localizedDescription = WEB_UI_STRING("Unable to parse manifest because of an unexpected format.", "WKWebExtensionErrorInvalidManifest description");
+        break;
+
+    case Error::UnsupportedManifestVersion:
+        localizedDescription = WEB_UI_STRING("An unsupported `manifest_version` was specified.", "WKWebExtensionErrorUnsupportedManifestVersion description");
+        break;
+
+    case Error::InvalidAction:
+        if (supportsManifestVersion(3))
+            localizedDescription = WEB_UI_STRING("Missing or empty `action` manifest entry.", "WKWebExtensionErrorInvalidManifestEntry description for action only");
+        else
+            localizedDescription = WEB_UI_STRING("Missing or empty `browser_action` or `page_action` manifest entry.", "WKWebExtensionErrorInvalidManifestEntry description for browser_action or page_action");
+        break;
+
+    case Error::InvalidActionIcon:
```

Comment:
You can put it here with a brace, like this:
```suggestion
    case Error::InvalidActionIcon: {
        RefPtr<JSON::Value> actionManifest;
```